leku/src/main/java/com/adevinta/leku/geocoder/GoogleGeocoderDataSource.kt

```diff
@@ -5,97 +5,143 @@ import com.google.android.gms.maps.model.LatLng
+import com.adevinta.leku.geocoder.api.SuggestionBuilder
 import java.util.Locale
 import org.json.JSONException
 
-private const val QUERY_REQUEST = "https://maps.googleapis.com/maps/api/geocode/json?address=%1\$s&key=%2\$s"
+private const val QUERY_REQUEST =
+    "https://maps.googleapis.com/maps/api/geocode/json?address=%1\$s&key=%2\$s"
 private const val QUERY_REQUEST_WITH_RECTANGLE =
     "https://maps.googleapis.com/maps/api/geocode/json?address=%1\$s&key=%2\$s&bounds=%3\$f,%4\$f|%5\$f,%6\$f"
-private const val QUERY_LAT_LONG = "https://maps.googleapis.com/maps/api/geocode/json?latlng=%1\$f,%2\$f&key=%3\$s"
+private const val QUERY_LAT_LONG =
+    "https://maps.googleapis.com/maps/api/geocode/json?latlng=%1\$f,%2\$f&key=%3\$s"
+private const val QUERY_AUTOCOMPLETE =
+    "https://maps.googleapis.com/maps/api/place/autocomplete/json?input=%1\$s&key=%2\$s"
+private const val QUERY_PLACE_ID =
+    "https://maps.googleapis.com/maps/api/place/details/json?place_id=%1\$s&key=%2\$s"
 
 class GoogleGeocoderDataSource(
     private val networkClient: NetworkClient,
-    private val addressBuilder: AddressBuilder
+    private val addressBuilder: AddressBuilder,
+    private val suggestionBuilder: SuggestionBuilder
 ) : GeocoderDataSourceInterface {
 
-    private var apiKey: String? = null
+    private var geolocationApiKey: String? = null
 
-    fun setApiKey(apiKey: String) {
-        this.apiKey = apiKey
+    fun setGeolocationApiKey(apiKey: String) {
+        this.geolocationApiKey = apiKey
+    }
+
+    private var placesApiKey: String? = null
+    fun setPlaceApiKey(apiKey: String) {
+        this.placesApiKey = apiKey
+    }
+
+    override fun autoCompleteFromLocationName(query: String): List<PlaceSuggestion> {
+        val suggestions = mutableListOf<PlaceSuggestion>()
+        if (placesApiKey == null) {
+            return suggestions
+        }
+        return try {
+            val result = networkClient.requestFromLocationName(
+                String.format(Locale.ENGLISH, QUERY_AUTOCOMPLETE, query, placesApiKey)
+            )
+            if (result != null) {
+                suggestions.addAll(suggestionBuilder.parseResult(result))
+            }
+            suggestions
+        } catch (e: JSONException) {
+            suggestions
+        } catch (e: NetworkException) {
+            suggestions
+        }
+    }
+
+    override fun getAddressFromPlaceId(placeId: String): Address? {
+        if (placesApiKey == null) {
+            return null
+        }
+        return try {
+            val result = networkClient.requestFromLocationName(
+                String.format(Locale.ENGLISH, QUERY_PLACE_ID, placeId, placesApiKey)
+            )
+            when {
+                result != null -> addressBuilder.parseResult(result)
+                else -> null
+            }
+        } catch (e: JSONException) {
+            null
+        } catch (e: NetworkException) {
+            null
```

Comment:
I'm not sure you have to hide this kind of error. Maybe the client wants to handle them.

Reply:
I kept the same `Exception` treatment with new data source classes: return an empty list or `null`